pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -62,41 +68,97 @@ func init() {
 }
 
 type CloudWatchService struct {
-	LogsService *LogsService `inject:""`
+	LogsService          *LogsService          `inject:""`
+	BackendPluginManager backendplugin.Manager `inject:""`
 }
 
 func (s *CloudWatchService) Init() error {
 	plog.Debug("initing")
 
-	tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
-		return newExecutor(s.LogsService), nil
+	im := datasource.NewInstanceManager(NewInstanceSettings())
+
+	factory := coreplugin.New(backend.ServeOpts{
+		QueryDataHandler: newExecutor(s.LogsService, im),
 	})
 
+	if err := s.BackendPluginManager.Register("cloudwatch", factory); err != nil {
+		plog.Error("Failed to register plugin", "error", err)
+	}
 	return nil
 }
 
-func newExecutor(logsService *LogsService) *cloudWatchExecutor {
+func newExecutor(logsService *LogsService, im instancemgmt.InstanceManager) *cloudWatchExecutor {
 	return &cloudWatchExecutor{
 		logsService: logsService,
+		im:          im,
+	}
+}
+
+func NewInstanceSettings() datasource.InstanceFactoryFunc {
+	return func(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		var jsonData map[string]string
+
+		err := json.Unmarshal(settings.JSONData, &jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+
+		model := datasourceInfo{
+			profile:       jsonData["profile"],
+			region:        jsonData["defaultRegion"],
+			assumeRoleARN: jsonData["assumeRoleArn"],
+			externalID:    jsonData["externalId"],
+			namespace:     jsonData["customMetricsNamespaces"],
+			datasourceID:  settings.ID,
+		}
+
+		atStr := jsonData["authType"]
+		at := authTypeDefault
+		switch atStr {
+		case "credentials":
+			at = authTypeSharedCreds
+		case "keys":
+			at = authTypeKeys
+		case "default":
+			at = authTypeDefault
+		case "arn":
+			at = authTypeDefault
+			plog.Warn("Authentication type \"arn\" is deprecated, falling back to default")
+		default:
+			plog.Warn("Unrecognized AWS authentication type", "type", atStr)
+		}
+
+		model.authType = at
+
+		if model.profile == "" {
+			model.profile = settings.Database // legacy support
+		}
+
+		model.accessKey = settings.DecryptedSecureJSONData["accessKey"]
+		model.secretKey = settings.DecryptedSecureJSONData["secretKey"]
+
+		return model, nil
 	}
 }
 
 // cloudWatchExecutor executes CloudWatch requests.
 type cloudWatchExecutor struct {
-	*models.DataSource
-
 	ec2Client  ec2iface.EC2API
 	rgtaClient resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
 
 	logsService *LogsService
+	im          instancemgmt.InstanceManager
 }
 
-func (e *cloudWatchExecutor) newSession(region string) (*session.Session, error) {
-	dsInfo := e.getDSInfo(region)
+func (e *cloudWatchExecutor) newSession(region string, pluginCtx backend.PluginContext) (*session.Session, error) {
+	dsInfo, err := e.getDSInfo(pluginCtx)
```

Comment:
It feels a bit odd to have a separate session cache RWMutex. But not easy to migrate away from I definitely understand. Something that we might look into in the future. 

Could potentially build upon the idea of this [instance manager](https://github.com/grafana/grafana-plugin-sdk-go/blob/master/backend/instancemgmt/instance_manager.go) as a way to replace the RWMutex with to remove lock-contention for cloudwatch data source instances?